### PR TITLE
IBP-4735 implement delete studies api

### PR DESCRIPTION
**Pull request by cuenyad**
Please review.

### CODE REVIEW

File: src/main/resources/messages_en.properties [hunk 1]
```diff
@@ -557,6 +557,9 @@ invalid.environment.data.id=Invalid environment data Id.
 invalid.variable.for.environment.data=The specified variable does not match for saved environment data.
 study.must.not.have.observation=The study must not have observation.
 methoddbid.invalid=Breeding Method DbId does not exist
+study.delete.not.permitted=You are not able to delete this study as you are not the owner. The owner is {0}.
+study.template.delete.not.permitted=Program templates cannot be deleted.
+study.id.not.exists=The study id {0} does not exist.
```
Comment (clarysabel) [inline src/main/resources/messages_en.properties:562]: Maybe reuse study.not.exist ?

File: src/main/java/org/ibp/api/java/user/UserService.java [hunk 2]
```diff
@@ -20,4 +20,13 @@
...
 
 	void updateUserProfile(final UserProfileUpdateRequestDTO userProfileUpdateRequestDTO, final WorkbenchUser workbenchUser);
+
+	/**
+	 * Gets the user by id.
+	 *
+	 * @param userId - the user id to match
+	 * @return the user matching the given id
+	 */
+	WorkbenchUser getUserById(final Integer userId);
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/java/user/UserService.java:30]: No not expose WorkbenchUser in this service, use UserDetailDto

File: src/main/java/org/ibp/api/java/impl/middleware/study/validator/StudyValidator.java [hunk 3]
```diff
@@ -117,18 +122,42 @@ public void validateStudyShouldNotHaveObservation(final Integer studyId) {
...
+		}
+
+		if (StringUtils.isBlank(study.getProgramUUID())) {
+			this.errors.reject("study.template.delete.not.permitted");
+			throw new ApiRequestValidationException(this.errors.getAllErrors());
+		}
+
+		final Integer studyUserId = study.getUser();
+		final WorkbenchUser user = this.securityService.getCurrentlyLoggedInUser();
+		if (studyUserId != null && !studyUserId.equals(user.getUserid())) {
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/java/impl/middleware/study/validator/StudyValidator.java:156]: If using UserDto, you can replace this if by !user.getUserId().equals(studyUserId)

File: src/main/java/org/ibp/api/java/impl/middleware/study/validator/StudyValidator.java [hunk 4]
```diff
@@ -117,18 +122,42 @@ public void validateStudyShouldNotHaveObservation(final Integer studyId) {
...
+		this.errors = new MapBindingResult(new HashMap<>(), Integer.class.getName());
 		if (this.studyService.studyHasGivenDatasetType(studyId, DatasetTypeEnum.MEANS_DATA.getId())) {
-			errors.reject("study.has.means.dataset");
-			throw new ApiRequestValidationException(errors.getAllErrors());
+			this.errors.reject("study.has.means.dataset");
+			throw new ApiRequestValidationException(this.errors.getAllErrors());
+		}
+	}
+
+	public void validateDeleteStudy(final Integer studyId) {
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/java/impl/middleware/study/validator/StudyValidator.java:140]: add tests

File: src/main/java/org/ibp/api/java/user/UserService.java (hunk 2 above)
Comment (cuenyad) [reply src/main/java/org/ibp/api/java/user/UserService.java]: done.

File: src/main/java/org/ibp/api/java/impl/middleware/study/validator/StudyValidator.java (hunk 3 above)
Comment (cuenyad) [reply src/main/java/org/ibp/api/java/impl/middleware/study/validator/StudyValidator.java]: done.

File: src/main/resources/messages_en.properties (hunk 1 above)
Comment (cuenyad) [reply src/main/resources/messages_en.properties:562]: The validatión in fieldbook return the id of the study.

File: src/main/java/org/ibp/api/java/impl/middleware/study/validator/StudyValidator.java (hunk 4 above)
Comment (cuenyad) [reply src/main/java/org/ibp/api/java/impl/middleware/study/validator/StudyValidator.java:140]: done.

**Review by clarysabel — CHANGES_REQUESTED** [review]
@cuenyad 
Leaving a ouple of comments in this PR

File: src/test/java/org/ibp/api/rest/dataset/validator/StudyValidatorTest.java [hunk 5]
```diff
@@ -228,4 +235,56 @@ public void testValidateStudyHasNoMeansDataset() {
 		}
 	}
 
+	@Test
+	public void testvalidateDeleteStudy_ThrowsException_WhenStudynotExist() {
```
[the remark this answers is not in the document]
Comment (clarysabel) [reply src/test/java/org/ibp/api/rest/dataset/validator/StudyValidatorTest.java:239]: WhenStudynotExist = WhenStudyNotExist

File: src/test/java/org/ibp/api/rest/dataset/validator/StudyValidatorTest.java [hunk 6]
```diff
@@ -228,4 +235,56 @@ public void testValidateStudyHasNoMeansDataset() {
...
+		try {
+			final Integer studyId = RandomUtils.nextInt();
+			this.studyValidator.validateDeleteStudy(studyId);
+		} catch (final ResourceNotFoundException e) {
+			assertThat(Arrays.asList(e.getError().getCodes()), hasItem("study.not.exist"));
+		}
+	}
+
+	@Test
+	public void testvalidateDeleteStudy_ThrowsException_WhenStudyHasProgramIUUD() {
```
Comment (clarysabel) [inline src/test/java/org/ibp/api/rest/dataset/validator/StudyValidatorTest.java:249]: fix camel case, 
fix typo IUUD = UUID

File: src/test/java/org/ibp/api/rest/dataset/validator/StudyValidatorTest.java [hunk 7]
```diff
@@ -228,4 +235,56 @@ public void testValidateStudyHasNoMeansDataset() {
...
+			study.setCreatedBy(USER_ID.toString());
+			Mockito.doReturn(study).when(studyDataManager).getStudy(studyId, false);
+			this.studyValidator.validateDeleteStudy(studyId);
+		} catch (final ApiRequestValidationException e) {
+			assertThat(Arrays.asList(e.getErrors().get(0).getCodes()), hasItem("study.template.delete.not.permitted"));
+		}
+	}
+
+	@Test
+	public void testvalidateDeleteStudy_ThrowsException_WhenTheUserIsNotOwnerOfStudy() {
```
Comment (clarysabel) [inline src/test/java/org/ibp/api/rest/dataset/validator/StudyValidatorTest.java:263]: fix camelCase

**Pull request merged**